### PR TITLE
Add no_sysex cargo feature for disabling system exclusive functionality

### PR DESCRIPTION
**Pull request by stuffmatic**
This PR adds the no_sysex cargo feature for disabling system exclusive functionality as discussed in https://github.com/AlexCharlton/midi-msg/issues/2.

* Using `no_sysex` currently results in a bunch of build warnings about unused functions in `time_code.rs` and `util.rs`. It feels like it should be possible to restructure the code a bit to allow easy exclusion of the unused functions.
* When a sysex message is encountered and `no_sysex` is used, I just return an error. An alternative approach could be to introduce an "ignored sysex" message type, which can be returned along with the number of bytes skipped.
* My first attempt at `no_sysex` support involved just adding `#[cfg...` statements where needed without restructuring the  existing code. There might be a cleaner way of doing this.

### CODE REVIEW

File: src/message.rs [hunk 1]
```diff
@@ -145,8 +149,12 @@ impl MidiMsg {
                 }
                 0xF => {
                     if b & 0b00001111 == 0 {
+                        #[cfg(not(feature = "no_sysex"))]
```
Comment (AlexCharlton) [inline src/message.rs:156]: I would separate out the logic here using the `cfg` macro and an `if` statement.

**Review by AlexCharlton — COMMENTED** [review]
> Using no_sysex currently results in a bunch of build warnings about unused functions in time_code.rs and util.rs. It feels like it should be possible to restructure the code a bit to allow easy exclusion of the unused functions.

That's a pain, but yeah, hopefully it won't be too hard to rearrange. Grouping into submodules that get conditionally re-exported should do the trick.

> When a sysex message is encountered and no_sysex is used, I just return an error. An alternative approach could be to introduce an "ignored sysex" message type, which can be returned along with the number of bytes skipped.

I think the error is a perfectly fine approach here.

> My first attempt at no_sysex support involved just adding #[cfg... statements where needed without restructuring the existing code. There might be a cleaner way of doing this.

In most places I think this is just fine, but I added one suggestion.

File: src/message.rs (hunk 1 above)
Comment (stuffmatic) [reply src/message.rs:156]: Not sure I'm following, sorry... However, I did realize you can disable entire blocks with the `cfg` macro like this, if you think that's acceptable

```
#[cfg(not(feature = "no_sysex"))]
 {
    let (msg, len) = SystemExclusiveMsg::from_midi(m, ctx)?;
    Ok((Self::SystemExclusive { msg }, len))
}
#[cfg(feature = "no_sysex")]
Err(ParseError::Invalid(format!("Got sysex message but the midi-msg crate was built with no_sysex.")))
```

Comment (AlexCharlton) [reply src/message.rs:156]: I was meaning something like this:

```
if cfg!(feature = "no_sysex") {
    Err(ParseError::Invalid(format!("Got sysex message but the midi-msg crate was built with no_sysex.")))
} else {
    let (msg, len) = SystemExclusiveMsg::from_midi(m, ctx)?;
    Ok((Self::SystemExclusive { msg }, len)
}
```

IMO, this makes the intent a little more clear: i.e. we have a true branch-point in behaviour.

Comment (stuffmatic) [reply src/message.rs:156]: Correct me if I'm wrong, but it seems like the `if cfg!`solution leads to both `if` branches being compiled and does not offer the conditional compilation we're after.

**Comment by stuffmatic** [comment]
> That's a pain, but yeah, hopefully it won't be too hard to rearrange. Grouping into submodules that get conditionally re-exported should do the trick.

I've moved some stuff from `util.rs` into `system_exclusive/util.rs` which took care of the build warnings related to `util.rs`. The only warnings left now are related to `time_code.rs`. Since you're more familiar with the spec and the code base, it would be great if you could provide some guidance about what parts of this file could be moved into the `system_exclusive` module.

File: src/message.rs (hunk 1 above)
Comment (AlexCharlton) [reply src/message.rs:156]: Using `cfg!` in a branching statement like that causes two things to happen when the code is compiled:

First macros are expanded, and `cfg!(...)` gets replaced with a boolean constant. Later the compiler backend removes branches that cannot be hit, so you wind up not paying a runtime cost (and thus it acts exactly like conditional compilation).

**Comment by AlexCharlton** [comment]
> I've moved some stuff from util.rs into system_exclusive/util.rs which took care of the build warnings related to util.rs. The only warnings left now are related to time_code.rs. Since you're more familiar with the spec and the code base, it would be great if you could provide some guidance about what parts of this file could be moved into the system_exclusive module.

Ah, I wasn't really suggesting creating a new `system_exclusive/file.rs`, but rather doing something like:

in `time_code.rs`:

```
#[cfg(not(feature = "no_sysex"))]
mod sysex_types {
  ...
}

#[cfg(not(feature = "no_sysex"))]
pub use sysex_types::*;
```

That said, `TimeCode` and the `TimeCodeType` should be the only types used outside of sysex.

File: src/message.rs (hunk 1 above)
Comment (stuffmatic) [reply src/message.rs:156]: Maybe I'm missing something obvious, but when I use `if cfg!(...` and run `cargo build` with the `no_sysex` feature enabled, the build fails with `failed to resolve: use of undeclared type SystemExclusiveMsg`.

Comment (AlexCharlton) [reply src/message.rs:156]: Nope, that's my bad. The branch removal doesn't happen until well after symbols are resolved, so it indeed can't be used like that if you're using a type that doesn't exist in one branch 🤦

**Comment by stuffmatic** [comment]
We're getting there :) I've moved sysex related code into conditionally exported modules in `util.rs` and `time_code.rs` like you suggested. Only the two warnings below remain. Should I mark these with `#[cfg(not(feature = "no_sysex"))]` for now?

```
warning: associated function is never used: `extend_midi`
  --> src/time_code.rs:36:19
   |
36 |     pub(crate) fn extend_midi(&self, v: &mut Vec<u8>) {
   |                   ^^^^^^^^^^^
   |
   = note: `#[warn(dead_code)]` on by default

warning: associated function is never used: `from_midi`
  --> src/time_code.rs:90:19
   |
90 |     pub(crate) fn from_midi(m: &[u8]) -> Result<Self, ParseError> {
   |                   ^^^^^^^^^
```

**Comment by AlexCharlton** [comment]
>  Should I mark these with #[cfg(not(feature = "no_sysex"))] for now?

Ah, yeah, I think that makes sense 👍

**Comment by AlexCharlton** [comment]
Nice stuff! Thanks for all the hard work!

**Pull request merged**